routes/x/module.tsx

```diff
@@ -350,59 +297,90 @@ function VersionSelector({
   );
 }
 
-export const handler: Handlers<MaybeData> = {
+export const handler: Handlers<DocPage | null> = {
   async GET(req, { params, render }) {
-    let {
-      name,
-      version,
-      path: maybePath,
-      symbol,
-    } = params as Params;
+    const { name, version, path } = params as Params;
     const url = new URL(req.url);
     const isHTML = accepts(req, "application/*", "text/html") === "text/html";
 
-    const path = maybePath ? "/" + maybePath : "";
-    const isStd = name === "std";
-
-    if (isStd && url.pathname.startsWith("/x")) {
+    if (name === "std" && url.pathname.startsWith("/x")) {
       url.pathname = url.pathname.slice(2);
       return Response.redirect(url, 301);
     }
 
-    if (!version) {
-      const versions = await getVersionList(name);
-      if (!versions?.latest) {
-        if (isHTML) {
-          return render!({ versions });
-        } else {
-          return new Response(
-            `The module '${name}' has no latest version`,
-            {
-              status: 404,
-              headers: {
-                "content-type": "text/plain",
-                "Access-Control-Allow-Origin": "*",
-              },
-            },
-          );
-        }
-      }
+    const symbol = url.searchParams.get("s");
+    const resURL = new URL(
+      `https://apiland.deno.dev/v2/modules/${name}/${
+        version || "__latest__"
+      }/page/${path}`,
+    );
+    if (symbol) {
+      resURL.searchParams.set("symbol", symbol);
+    }
+
+    let data: DocPage;
 
+    const res = await fetch(resURL, {
+      redirect: "manual",
+    });
+    if (res.status === 404) { // module doesnt exist
+      if (isHTML) {
+        return render(null);
+      } else {
+        return new Response(`The module '${name}' does not exist`, {
+          status: 404,
+        });
+      }
+    } else if (res.status === 302) { // implicit latest
+      const latestVersion = res.headers.get("X-Deno-Latest-Version")!;
       return new Response(undefined, {
         headers: {
-          Location: getModulePath(name, versions!.latest, path),
-          "x-deno-warning": `Implicitly using latest version (${
-            versions!.latest
-          }) for ${url.href}`,
+          Location: getModulePath(
+            name,
+            latestVersion,
+            path ? ("/" + path) : undefined,
+          ),
+          "x-deno-warning":
+            `Implicitly using latest version (${latestVersion}) for ${url.href}`,
           "Access-Control-Allow-Origin": "*",
         },
         status: 302,
       });
+    } else if (res.status === 301) { // path is directory and there is an index module
+      const newPath = res.headers.get("X-Deno-Module-Path")!;
+      return new Response(undefined, {
+        headers: {
+          Location: getModulePath(
+            name,
+            version,
+            newPath,
+          ),
+        },
+        status: 301,
+      });
```

Comment:
one "nasty" aspect about this is that it also affects imports, so importing `https://deno.land/std@0.150.0/collections` will work as it redirects to `https://deno.land/std@0.150.0/collections/mod.ts`. i will fix this.

Reply:
should we respond with an error for raw as a directory is being requested?

Reply:
for raws, it should be a 404 I think, or a 406 Not Acceptable (which means we can't negotiate content to provide a response, which is actually more accurate)